pkg/vault/contracts/VaultCommon.sol

```diff
@@ -300,22 +309,14 @@ abstract contract VaultCommon is IVaultEvents, IVaultErrors, VaultStorage, Reent
 
     /**
      * @dev Fill in PoolData, including paying protocol yield fees and computing final raw and live balances.
-     * This function modifies protocol fees and balance storage. Since it modifies storage and makes external
-     * calls, it must be nonReentrant.
      * Side effects: updates `_aggregateFeeAmounts` and `_poolTokenBalances` in storage.
      */
     function _loadPoolDataUpdatingBalancesAndYieldFees(
         address pool,
         Rounding roundingDirection
-    ) internal nonReentrant returns (PoolData memory poolData) {
```

Comment:
In `poolData.load`, we update rates with `getTokenRate()`, which calls `getRate()` from the RateProvider. Although getRate is a view function, someone can manipulate the rate somehow. For example, swap() calls this function, and when getting the rate, the attacker executes a different vault function (like addLiquidity) with the manipulated rate and then returns another rate in the swap. Not sure if this is entirely possible, but it'd be a Read-only reentrancy.

Swap, Add and Remove liquidity external calls are reentrant, so should we remove the nonReentrant modifier from here? I mean, the attack above would only happen with a malicious rate provider, but with permissionless rate providers, maybe the nonReentrant modifier here is a guardrail.

Reply:
The auditors brought this up. If it's actually unsafe, then instead of removing it, we'd move it to `_loadPoolData`, to protect both.

Reply:
Got it.

Reply:
Would like to hear what @jubeira / the auditors think before merging.

Reply:
you can only call `view` functions anyways.

If the rate provider happens to call the vault to get some data, that could be a problem. But I don't think that's a valid use-case in principle.

> If it's actually unsafe, then instead of removing it, we'd move it to _loadPoolData, to protect both.

I'd wait a bit longer before merging; this is true.